src/tools/plan-generator.ts

```diff
@@ -0,0 +1,148 @@
+      this.fileHeader(),
+      (await this.createPlans()).join('\n'),
+      this.fileFooter()
+    ];
+
+    return planOutline.join('\n');
+  }
+
+  /** Converts a resolved recipe into a `Plan` object. */
+  async createPlans(): Promise<string[]> {
+    const plans = [];
+    for (const recipe of this.resolutions) {
+      const planName = `${recipe.name.replace(/[rR]ecipe/, '')}Plan`;
+
+      const particles = [];
+      for (const particle of recipe.particles) {
+        particles.push(await this.createParticle(particle));
+      }
+
+      const start = `object ${planName} : `;
+      const plan = `${start}${ktUtils.applyFun('Plan', particles, 'Plan', start.length)}`;
+      plans.push(plan);
+    }
+    return plans;
+  }
+
+  /** Generates a Kotlin `Plan.Particle` instantiation from a Particle. */
+  async createParticle(particle: Particle): Promise<string> {
+    const spec = particle.spec;
+    const location = (spec && (spec.implBlobUrl || (spec.implFile && spec.implFile.replace('/', '.')))) || '';
+
+    const particleName = `"${particle.name}"`;
+    const locationArg = `"${location}"`;
+
+    const connectionMappings = [];
+    for (const [key, conn] of Object.entries(particle.connections)) {
+      connectionMappings.push(`"${key}" to ${await this.createHandleConnection(conn)}`);
+    }
+
+    return ktUtils.applyFun('Particle', [particleName, locationArg, ktUtils.mapOf(connectionMappings)]);
+  }
+
+  /** Generates a Kotlin `Plan.HandleConnection` from a HandleConnection. */
+  async createHandleConnection(connection: HandleConnection): Promise<string> {
+    const storageKey = this.createStorageKey(connection.handle.storageKey);
+    let mode;
+    switch (connection.direction) {
+      case 'reads':
+        mode = 'HandleMode.Read';
+        break;
+      case 'writes':
+        mode = 'HandleMode.Write';
+        break;
+      case 'reads writes':
+        mode = 'HandleMode.ReadWrite';
+        break;
+      default:
+        throw new PlanGeneratorError(`HandleConnection direction '${connection.direction}' is not supported.`);
+    }
+    const type = await this.createType(connection.type);
+    const ttl = 'null';
+
+    return ktUtils.applyFun('HandleConnection', [storageKey, mode, type, ttl]);
+  }
+
+  /** Generates a Kotlin `StorageKey` from a StorageKey. */
+  createStorageKey(storageKey: StorageKey | undefined): string {
+    return `StorageKeyParser.parse("${(storageKey || '').toString()}")`;
+  }
+
+  /** Generates a Kotlin `core.arc.type.Type` from a Type. */
+  async createType(type: Type): Promise<string> {
+    switch (type.tag) {
+      case 'Collection':
+        return ktUtils.applyFun('CollectionType', [await this.createType(type.getContainedType())]);
+      case 'Count':
+        return ktUtils.applyFun('CountType', [await this.createType(type.getContainedType())]);
+      case 'Entity':
+        return ktUtils.applyFun('EntityType', [`SchemaRegistry["${await type.getEntitySchema().hash()}"]`])
```

Comment:
I don't think you should use the hash, because the hash is lazily initialized. If for example, someone refers to the Plan's HandleConnection.type field prior to ever referencing EntitySpec.schema, it won't get registered and the registry will return null. 

You should probably just insert a direct call to schema()